cpp/ql/src/experimental/Security/CWE/CWE-409-DecompressionBomb/DecompressionBombs.ql

```diff
@@ -0,0 +1,186 @@
+/**
+ * @name User-controlled file decompression
+ * @description User-controlled data that flows into decompression library APIs without checking the compression rate is dangerous
+ * @kind path-problem
+ * @problem.severity error
+ * @security-severity 7.8
+ * @precision high
+ * @id cpp/data-decompression
+ * @tags security
+ *       experimental
+ *       external/cwe/cwe-409
+ */
+
+import cpp
+import semmle.code.cpp.ir.dataflow.TaintTracking
+import semmle.code.cpp.security.FlowSources
+import semmle.code.cpp.commons.File
+import Bzip2
+import Brotli
+import LibArchive
+import LibMiniz
+import ZSTD
+import MiniZip
+import XZ
+import ZlibGzopen
+import ZlibUncompress
+import ZlibInflator
+import Brotli
+
+module DecompressionTaintConfig implements DataFlow::ConfigSig {
+  predicate isSource(DataFlow::Node source) { source instanceof FlowSource }
+
+  predicate isSink(DataFlow::Node sink) {
+    exists(FunctionCall fc | fc.getTarget() instanceof BrotliDecoderDecompressStreamFunction |
+      fc.getArgument(2) = sink.asExpr()
+    )
+    or
```

Comment:
This sink is rather difficult to read. It would be better to define an abstract class, like:
```ql
abstract class DecompressionFunction extends Function {
  abstract int getArchiveParameterIndex();
}
```
and then have the functions that are used as sinks extend that abstract class, for example:
```ql
class BZ2BzDecompressFunction extends DecompressionFunction {
  BZ2BzDecompressFunction() { this.hasGlobalName(["BZ2_bzDecompress"]) }

  override int getArchiveParameterIndex() { result = 0 }
}
```

Then the sink can be written concisely as:
```ql
predicate isSink(DataFlow::Node sink) {
  exists(FunctionCall fc, DecompressionFunction f | fc.getTarget() = f |
    fc.getArgument(f.getArchiveParameterIndex()) = sink.asExpr()
  )
}
```